Trim down JS

michaelcontento opened the pull request:
Hi,

on my personal blog I don't need momentJS and also the builtin syntax highlighting of hugo is totally enough for me, thus I also don't highlightJS on my page.

This is especially annoying, as those two are quite big dependencies after all 😢 

So in this PR I've managed to create multiple variations of the `theme.js` file wether or not it contains momentJS, highlightJS, both or none of them. With those varied files at hand (all managed by laravel-mix so nothing to maintain manually btw!) we can simply decide in our `baseof.html` which one to load.

In order to keep a somewhat consistent look & feel I also picked `pastie` as the theme of the builtin syntax highlighter of hugo.

Here the results

| `enableHighlightJs` | `enableMomentJs` | output | size | size gzipped |
|----|---|--|--|--|
| ✅ | ✅ | `theme.js` | 1.6MiB | 474KiB |
| ✅ | ❌ | `theme-exclude-moment.js` | 1.09MiB | 347KiB |
| ❌ | ✅ | `theme-exclude-highlight.js` | 806KiB | 212KiB |
| ❌ | ❌ | `theme-exclude-moment-and-highlight.js` | 288KiB | 85KiB |

So, from a performance point of view, I think this is a big win for some users 😄

Lednerb commented:
This is a nice addition and can indeed speed-up loading times!

I'd like to ask you to add an additional section in the `README.md` for documentation purposes.

Also this would force us to do a version increase because it's not backwards compatible at the moment and will likely break existing installations if the users will update their theme version.


However, I like the approach to `enable` the features that user's want instead of `disabling` not wanted features.
@igor-baiborodine what do you think about this PR?

michaelcontento commented:
Oh! My fault, both features should have been ENABLED by default. I‘ll fix that when I‘m back at my desk. 

But the backwards incompatibility got me thinking. I could also change around the logic behind the filenames! 

So that, in the end, the theme.js would contain anything and both features would be enabled by default. This would relief us (and all users) from the breaking change. 

It would simply become a new flag that existing users could use if they want 😀

Maybe in the future there are some other reasons for a major release where we could change the default values to false. But IMHO we are not required to do so now. 

What do you think? Change it or leave it as it is?


And regarding the README: I‘ll add a section later today 👍

michaelcontento commented:
So .. I updated this PR once again 😄 

Now we should be 100% backwards compatible (config & JS filename wise) and I've added a section to the `README.md` 🚀

igor-baiborodine commented:
@Lednerb, In March, I was planning to start working on the `v4` version that will include the following breaking changes:
1. Migrate to Hugo's built-in menus for the header (https://github.com/Lednerb/bilberry-hugo-theme/discussions/436)
2. Migrate to Hugo pipelines for SCCC processing (https://github.com/Lednerb/bilberry-hugo-theme/discussions/339)

Also, I wanted to include the migration to FontAwesome v6 (https://github.com/Lednerb/bilberry-hugo-theme/discussions/358).

Therefore, for this PR, may I suggest putting it on hold, so it can be incorporated later into `v4` with other breaking changes?

michaelcontento commented:
Changes to the code base are, IMHO, quite small and the latest version of the PR preserves full backwards compatibility and introduces no breaking change. 

So from my point of view there is no real need to postpone this 🤔

igor-baiborodine commented:
@michaelcontento, if the backward compatibility is preserved, it should be fine to merge into the `v3`.

michaelcontento commented:
Yes, it is preserved as far as I can tell 😊 

- ✅ Default values for the settings match the old behaviour
- ✅ The newly introduced setting, `enableMomentJs`, is also properly handled if not set at all
- ✅ Content of the `theme.js` has not been changed, so we even support users that have hardcoded references in a modified theme are safe 😄

igor-baiborodine commented on README.md:
```diff
@@ -926,3 +939,6 @@ of any kind welcome!
 
 The Bilberry Hugo theme is licensed under the MIT license.
+
+  [highlightjs]: https://highlightjs.org/
```
To stay uniform, all links in README should be added using the inline syntax, i.e., \[link text\]\(URL\).

igor-baiborodine commented on webpack.mix.js:
```diff
@@ -17,7 +17,12 @@ mix.autoload({
-.js('assets/js/theme.js', './')
+// MomentJS and HightlighJS are quire big and thus we create variants
+// that can be used with or without them!
+.js('assets/js/theme.js', './theme-noMoment-noHighlight.js')
```
Please no `camelCase` in file names, therefore:
1. `theme-no-moment-and-highlight.js` or `theme-exclude-moment-and-highlight.js`
2. `theme-no-moment.js` or `theme-exclude-moment.js`
3. `theme-no-highlight.js` or `theme-exclude-highlight.js`

igor-baiborodine commented on assets/sass/_bilberry-hugo-theme.scss:
```diff
@@ -31,9 +31,13 @@
         }
     }
 
+    code {
```
Why is this change needed? How can I test it?

igor-baiborodine commented:
@michaelcontento, your PR is missing the `static/theme.css`. It should be include like in this PR: https://github.com/Lednerb/bilberry-hugo-theme/pull/223/files

![image](https://user-images.githubusercontent.com/1027701/218193432-b45b88f7-b0e6-476d-9ee5-1bf35b91a04f.png)

igor-baiborodine commented:
@michaelcontento, thanks for your contribution!

Your proposed changes were merged to the `master` branch via a new branch except for the ones below. Therefore, I'm closing this PR. 

![image](https://user-images.githubusercontent.com/1027701/220130695-f1d62334-f394-4256-8870-7d90d6f5e51c.png)

![image](https://user-images.githubusercontent.com/1027701/220130853-a0a482f8-58e3-44a4-b8c3-74e8f2327756.png)
